README
====
iiif
====

.. image:: https://travis-ci.org/zimeon/iiif.png?branch=main
  :target: https://travis-ci.org/zimeon/iiif?branch=main
  :alt: Build status

.. image:: https://coveralls.io/repos/zimeon/iiif/badge.png?branch=main
  :target: https://coveralls.io/r/zimeon/iiif?branch=main
  :alt: Test coverage

*iiif* - IIIF Image API reference implementation and Python library.
Supports the `International Image Interoperability Framework
<http://iiif.io/>`_: Image API
`2.1
<http://iiif.io/api/image/2.1/>`_ and
`3.0
<http://iiif.io/api/image/3.0/>`_ (as well as legacy versions
`2.0
<http://iiif.io/api/image/2.0/>`_,
`1.1
<http://iiif.io/api/image/1.1/>`_,
`1.0
<http://iiif.io/api/image/1.0/>`_).
Also includes a test server using the library to implement the Image
API transformations dynamically, and a static file generator to
pre-compute tiles necessary for OpenSeadragon which may then be
stored a static files.

Installation
------------

The library, test server, static file generator are all designed to
work with recent versions of Python 3 and Python 2.7. Manual installation
is necessary to get the demonstration documentation and examples.

**Automatic installation from PyPI**

The *iiif* library code and scripts are available from `PyPI
<http://pypi.python.org/pypi/iiif>`_ and can be installed with
``pip``, e.g.::

    pip install iiif

The PyPI version does not include all demonstration documentation and
examples.

**Manual installation from github**

First, clone the code and examples from the `Github repository
<https://github.com/zimeon/iiif>`_::

    cd /tmp
    git clone git://github.com/zimeon/iiif.git
    cd iiif/

Second, install ``Pillow`` (a fork of ``PIL``)::

    pip install Pillow

You may need ``sudo`` if installing system-wide as opposed to in a user-space
environment.

After installation of ``Pillow`` there will be be a summary of image formats
supported. Addition libraries may be require to get all formats depending
on your platform.

The scripts and demonstrations can then be run from the `iiif` directory.

Finally, you may install the library code and scripts locally in the
appropriate places within your python setup using::

    python setup.py build
    python setup.py install

- You may need ``sudo`` to if installing system-wide as opposed to in a user-space environment

The source code is maintained on `Github
<https://github.com/zimeon/iiif>`_

Contributing
------------

Bug reports welcome as `github issues
<https://github.com/zimeon/iiif/issues>`_.

See `CONTRIBUTING.md
<https://github.com/zimeon/iiif/blob/main/CONTRIBUTING.md>`_
for guidelines for contributing, and the `list of contributors
<https://github.com/zimeon/iiif/graphs/contributors>`_

See also
--------

- https://github.com/zimeon/iiif -- git repository
- https://github.com/zimeon/iiif/tree/main/demo-static -- Static File Demo with OpenSeadragon
- https://github.com/zimeon/iiif/tree/main/demo-auth -- IIIF Authentication API Demo

Copyright and License
---------------------

    iiif library and programs implementing the IIIF API
    Copyright (C) 2012--2021 Simeon Warner

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

See `LICENSE.txt
<LICENSE.txt>`_


CHANGES.md
# iiif changelog

2021-05-05 v1.0.10

  * Add scheme to config to support https (https://github.com/zimeon/iiif/pull/52)
  * Add python 3.9 to testing

2020-04-16 v1.0.9

  * No code changes
  * Drop testing support for python 3.4, add python 3.7 and 3.8
  * Tweak Travis build to keep python 2.7 tests happy

2019-11-09 v1.0.8

  * Removed test image `testimages/test1_224x167.png` (#45)
  * Add helpful error message from wrong use of size full in 3.0, max in pre 2.1 (#43)
  * Add check and helpful error message for use of size with caret pre 3.0 (#48)

2019-08-21 v1.0.7

  * Add preliminary support for IIIF Image API v3.0 in iiif_testserver.py (#38)
  * Drop support for python 3.4

2018-03-05 v1.0.6

  * Use latest version of Pillow (currently 5.0.0)
  * Drop support for python 2.6

2018-02-16 v1.0.5

  * Refactor most of Flask code from iiif_testserver.py into iiif.flask_utils
  * Create iiif_reference_server.py for use with http://iiif.io/ website
  * Allow use of config files for iiif_testserver.py, iiif_reference_server.py through change from argparse to ConfigArgParser

2017-04-20 v1.0.4

  * Add External authentication interaction pattern

2017-03-20 v1.0.3

  * Add --extra parameter to iiif_static.py
  * Drop python 3.3 from tests, add python 3.6
  * Drop rst-lint from travis tests as no longer works on python 2.6

2017-01-19 v1.0.2

  * Handle 16bit TIFF with PIL manipulators
  * Support IIIF Authentication API v1.0

2016-05-24 v1.0.1

  * Tweak pypi install

2016-05-24 v1.0.0

  * Change to 1.x.x version as core codebase stable
  * Make IIIF Image API v2.1 the default (although not all features implemented)
  * Improvements for PIL manipulator (thanks @jgreidy)
  * Extended info.profile to complex profiles for 2.x Image Information 

2016-04-22 v0.6.1

  * Now works with python 2.6, 2.7, 3.3, 3.4 and 3.5
  * Static tile generation supports both use of canonical size syntax for
    OpenSeadragon >= 1.2.1, and the old syntax for earlier versions
  * Better surface/handle image size warnings from Pillow
  * Use logging instead of print in iiif.static

2015-08-17 v0.6.0

  * Refactor manipulators for easier testing
  * Improve test coverage of PIL manipulator
  * Fix static files tile sizes (thanks @edsu)
  * Modify static tile generation for canonical URIs as used by OpenSeadragon 2.0
  * Test server now a Flask application
  * DRAFT - IIIF Image API v2.1 features for testing (not final), add /square/
    region, and trial authentication support

2015-02-14 v0.5.1

  * Valentines edition, wishing Amy and Ian lasting happiness
  * PIL and netpbm manipulators support IIIF Image API v2.0 and v1.1 at level 2
  * Improved test coverage using IIIF validator
    (https://pypi.python.org/pypi/iiif-validator)

2014-11-11 v0.5.0

  * Supports IIIF Image API v2.0 and v1.1 at level 1

2014-09-17 v0.4.2

  * Tidy tests, checked against python 2.6 and 2.7
  * Still needs work to support recently released IIIF Image API v2.0

2014-04-28 v0.4.1

  * Add iiif_static.py as script for pypi install

2014-04-28 v0.4.0

  * Aim to support IIIF Image API v1.1 and v2.0
  * Added generation of static file tiles for OpenSeadragon
  * Included demo for OpenSeadragon with static file tiles

2014-04-22 v0.3.0

  * Change pypi package name from i3f to iiif

2013-05-21 v0.2.0

  * Reorganized, aims for IIIF Image API v1.0
    (http://www-sul.stanford.edu/iiif/image-api/)
  * PIL library support (IIIFManipulatorPIL) and null tested, the netpbm
    version (IIIFManipulatorNetpbm) has not been updated/tested and is
    likely broken.
  * Designed to work with Loris for for Open Seadragon demo
    (https://github.com/zimeon/loris)
  * Add GPL

2012-03-21 v0.1.0

  * First stab at IIIF Image API v0.1


iiif/static.py
"""Static file generation for IIIF Image API.

Use IIIF Image API manipulations to generate a set of tiles for
a level0 implementation of the IIIF Image API using static files.
"""

import math
import logging
import os
import os.path
import shutil
from string import Template

from .manipulator_pil import IIIFManipulatorPIL
from .manipulator_gen import IIIFManipulatorGen
from .info import IIIFInfo
from .request import IIIFRequest
from .error import IIIFZeroSizeError


def static_partial_tile_sizes(width, height, tilesize, scale_factors):
    """Generator for partial tile sizes for zoomed in views.

    Positional arguments:
    width -- width of full size image
    height -- height of full size image
    tilesize -- width and height of tiles
    scale_factors -- iterable of scale factors, typically [1,2,4..]

    Yields ([rx,ry,rw,rh],[sw,sh]), the region and size for each tile
    """
    for sf in scale_factors:
        if (sf * tilesize >= width and sf * tilesize >= height):
            continue  # avoid any full-region tiles
        rts = tilesize * sf  # tile size in original region
        xt = (width - 1) // rts + 1
        yt = (height - 1) // rts + 1
        for nx in range(xt):
            rx = nx * rts
            rxe = rx + rts
            if (rxe > width):
                rxe = width
            rw = rxe - rx
            # same as sw = int(math.ceil(rw/float(sf)))
            sw = (rw + sf - 1) // sf
            for ny in range(yt):
                ry = ny * rts
                rye = ry + rts
                if (rye > height):
                    rye = height
                rh = rye - ry
                # same as sh = int(math.ceil(rh/float(sf)))
                sh = (rh + sf - 1) // sf
                yield([rx, ry, rw, rh], [sw, sh])


def static_full_sizes(width, height, tilesize):
    """Generator for scaled-down full image sizes.

    Positional arguments:
    width -- width of full size image
    height -- height of full size image
    tilesize -- width and height of tiles

    Yields [sw,sh], the size for each full-region tile that is less than
    the tilesize. This includes tiles up to the full image size if that
    is smaller than the tilesize.
    """
    # FIXME - Not sure what correct algorithm is for this, from
    # observation of Openseadragon it seems that one keeps halving
    # the pixel size of the full image until until both width and
    # height are less than the tile size. After that all subsequent
    # halving of the image size are used, all the way down to 1,1.
    # It seems that without these reduced size full-region images,
    # OpenSeadragon will not display any unzoomed image in small windows.
    #
    # I do not understand the algorithm that OpenSeadragon uses (or
    # know where it is in the code) to decide how small a version of
    # the complete image to request. It seems that there is a bug in
    # OpenSeadragon here because in some cases it requests images
    # of size 1,1 multiple times, which is anyway a useless image.
    for level in range(0, 20):
        factor = 2.0**level
        sw = int(width / factor + 0.5)
        sh = int(height / factor + 0.5)
        if (sw < tilesize and sh < tilesize):
            if (sw < 1 or sh < 1):
                break
            yield([sw, sh])


class IIIFStaticError(Exception):
    """Error class for errors to be resported to user."""

    pass


class IIIFStatic(object):
    """Provide static generation of IIIF images.

    Simplest, using source image as model for directory which
    will be in same directory without extension:

        IIIFStatic("image1.jpg").generate()

    More complex with different output directory but using name
    derived from image:

        sg = IIIFStatic(dst="outdir")
        sg.generate("image2.jpg")
        sg.generate("image3.jpg")

    The class is quite noisy at level logging.INFO, set the logging
    level to logging.WARNING to get log output only when there are
    warnings or errors.
    """

    def __init__(self, src=None, dst=None, tilesize=None,
                 api_version='2.0', dryrun=None, prefix='',
                 osd_version=None, generator=False,
                 max_image_pixels=0, extras=[]):
        """Initialization for IIIFStatic instances.

        All keyword arguments are optional:
        src -- source image file
        dst -- destination directory
        tilesize -- size of square tiles to generate (default 512)
        api_version -- IIIF Image API version to support (default 2.0)
        dryrun -- True to not write any output (default None)
        prefix -- identifier prefix
        osd_version -- use a specific version of OpenSeadragon
        extras -- extras request parameters to generate for
        """
        self.src = src
        self.dst = dst
        self.tilesize = tilesize if tilesize is not None else 512
        self.api_version = api_version
        if (self.api_version == '1'):
            self.api_version = '1.1'
        elif (self.api_version == '2'):
            self.api_version = '2.0'
        self.dryrun = (dryrun is not None)
        self.logger = logging.getLogger(__name__)
        self.prefix = prefix
        self.osd_version = osd_version if osd_version else '2.0.0'
        if (generator):
            self.manipulator_klass = IIIFManipulatorGen
        else:
            self.manipulator_klass = IIIFManipulatorPIL
        self.max_image_pixels = max_image_pixels
        # parse values in extras before adding to list, remove any leading /
        # if present on extras values
        self.extras = []
        for extra in extras:
            if extra.startswith('/'):
                extra = extra[1:]
            r = IIIFRequest(identifier='dummy',
                            api_version=self.api_version)
            r.parse_url(extra)
            self.extras.append(r)
        # config for locations of OpenSeadragon
        # - dir is relative to base, will be copied to dir under html_dir
        # - js and images are relative to dir
        self.module_dir = os.path.dirname(__file__)
        self.osd_config = {
            '1.0.0': {
                'base': os.path.join(self.module_dir, 'third_party'),
                'dir': 'openseadragon100',
                'js': 'openseadragon.min.js',
                'images': 'images',
                'use_canonical': False,
                'notes': 'Uses /w,h/ for size, algorithm not quite right, API v1.1'
            },
            '1.2.1': {
                'base': os.path.join(self.module_dir, 'third_party'),
                'dir': 'openseadragon121',
                'js': 'openseadragon.min.js',
                'images': 'images',
                'use_canonical': True,
                'notes': "Uses /w,/ canonical syntax, works with API v1.1,2.0"
            },
            '2.0.0': {
                'base': os.path.join(self.module_dir, 'third_party'),
                'dir': 'openseadragon200',
                'js': 'openseadragon.min.js',
                'images': 'images',
                'use_canonical': True,
                'notes': "Uses /w,/ canonical syntax, works with API v1.1,2.0"
            }
        }
        # used internally:
        self.identifier = None
        self.copied_osd = False
        self.template_dir = os.path.join(self.module_dir, 'templates')

    def get_osd_config(self, osd_version):
        """Select appropriate portion of config.

        If the version requested is not supported the raise an exception with
        a helpful error message listing the versions supported.
        """
        if (osd_version in self.osd_config):
            return(self.osd_config[osd_version])
        else:
            raise IIIFStaticError("OpenSeadragon version %s not supported, available versions are %s" %
                                  (osd_version, ', '.join(sorted(self.osd_config.keys()))))

    def generate(self, src=None, identifier=None):
        """Generate static files for one source image."""
        self.src = src
        self.identifier = identifier
        # Get image details and calculate tiles
        im = self.manipulator_klass()
        im.srcfile = self.src
        im.set_max_image_pixels(self.max_image_pixels)
        im.do_first()
        width = im.width
        height = im.height
        scale_factors = im.scale_factors(self.tilesize)
        # Setup destination and IIIF identifier
        self.setup_destination()
        # Write out images
        for (region, size) in static_partial_tile_sizes(width, height, self.tilesize, scale_factors):
            self.generate_tile(region, size)
        sizes = []
        for size in static_full_sizes(width, height, self.tilesize):
            # See https://github.com/zimeon/iiif/issues/9
            sizes.append({'width': size[0], 'height': size[1]})
            self.generate_tile('full', size)
        for request in self.extras:
            request.identifier = self.identifier
            if (request.region_full and
                    request.size_wh[0] is not None and
                    request.size_wh[1] is not None and
                    request.rotation_deg == 0.0 and
                    request.quality == request.default_quality and
                    request.format == 'jpg'):
                # FIXME - should this test of be pushed into iiif.request?
                sizes.append({'width': request.size_wh[0], 'height': request.size_wh[1]})
            self.generate_file(request)
        # Write info.json
        qualities = ['default'] if (self.api_version > '1.1') else ['native']
        info = IIIFInfo(level=0, server_and_prefix=self.prefix, identifier=self.identifier,
                        width=width, height=height, scale_factors=scale_factors,
                        tile_width=self.tilesize, tile_height=self.tilesize,
                        formats=['jpg'], qualities=qualities, sizes=sizes,
                        api_version=self.api_version)
        json_file = os.path.join(self.dst, self.identifier, 'info.json')
        if (self.dryrun):
            self.logger.warning(
                "dryrun mode, would write the following files:")
            self.logger.warning("%s / %s/%s" %
                                (self.dst, self.identifier, 'info.json'))
        else:
            with open(json_file, 'w') as f:
                f.write(info.as_json())
                f.close()
            self.logger.info("%s / %s/%s" %
                             (self.dst, self.identifier, 'info.json'))
            self.logger.debug("Written %s" % (json_file))

    def generate_tile(self, region, size):
        """Generate one tile for this given region, size of this image."""
        r = IIIFRequest(identifier=self.identifier,
                        api_version=self.api_version)
        if (region == 'full'):
            r.region_full = True
        else:
            r.region_xywh = region  # [rx,ry,rw,rh]
        r.size_wh = size  # [sw,sh]
        r.format = 'jpg'
        self.generate_file(r, True)

    def generate_file(self, r, undistorted=False):
        """Generate file for IIIFRequest object r from this image.

        FIXME - Would be nicer to have the test for an undistorted image request
        based on the IIIFRequest object, and then know whether to apply canonicalization
        or not.

        Logically we might use `w,h` instead of the Image API v2.0 canonical
        form `w,` if the api_version is 1.x. However, OSD 1.2.1 and 2.x assume
        the new canonical form even in the case where the API version is declared
        earlier. Thus, determine whether to use the canonical or `w,h` form based
        solely on the setting of osd_version.
        """
        use_canonical = self.get_osd_config(self.osd_version)['use_canonical']
        if (undistorted and use_canonical):
            height = r.size_wh[1]
            r.size_wh = [r.size_wh[0], None]  # [sw,sh] -> [sw,]
        path = r.url()
        # Generate...
        if (self.dryrun):
            self.logger.info("%s / %s" % (self.dst, path))
        else:
            m = self.manipulator_klass(api_version=self.api_version)
            try:
                m.derive(srcfile=self.src, request=r,
                         outfile=os.path.join(self.dst, path))
                self.logger.info("%s / %s" % (self.dst, path))
            except IIIFZeroSizeError:
                self.logger.info("%s / %s - zero size, skipped" %
                                 (self.dst, path))
                return  # done if zero size
        if (r.region_full and use_canonical):
            # In v2.0 of the spec, the canonical URI form `w,` for scaled
            # images of the full region was introduced. This is somewhat at
            # odds with the requirement for `w,h` specified in `sizes` to
            # be available, and has problems of precision with tall narrow
            # images. Hopefully will be fixed in 3.0 but for now symlink
            # the `w,h` form to the `w,` dirs so that might use the specified
            # `w,h` also work. See
            # <https://github.com/IIIF/iiif.io/issues/544>
            #
            # FIXME - This is ugly because we duplicate code in
            # iiif.request.url to construct the partial URL
            region_dir = os.path.join(r.quote(r.identifier), "full")
            wh_dir = "%d,%d" % (r.size_wh[0], height)
            wh_path = os.path.join(region_dir, wh_dir)
            wc_dir = "%d," % (r.size_wh[0])
            wc_path = os.path.join(region_dir, wc_dir)
            if (not self.dryrun):
                ln = os.path.join(self.dst, wh_path)
                if (os.path.exists(ln)):
                    os.remove(ln)
                os.symlink(wc_dir, ln)
            self.logger.info("%s / %s -> %s" % (self.dst, wh_path, wc_path))

    def setup_destination(self):
        """Setup output directory based on self.dst and self.identifier.

        Returns the output directory name on success, raises and exception on
        failure.
        """
        # Do we have a separate identifier?
        if (not self.identifier):
            # No separate identifier specified, split off the last path segment
            # of the source name, strip the extension to get the identifier
            self.identifier = os.path.splitext(os.path.split(self.src)[1])[0]
        # Done if dryrun, else setup self.dst first
        if (self.dryrun):
            return
        if (not self.dst):
            raise IIIFStaticError("No destination directory specified!")
        dst = self.dst
        if (os.path.isdir(dst)):
            # Exists, OK
            pass
        elif (os.path.isfile(dst)):
            raise IIIFStaticError(
                "Can't write to directory %s: a file of that name exists" % dst)
        else:
            os.makedirs(dst)
        # Second, create identifier based subdir if necessary
        outd = os.path.join(dst, self.identifier)
        if (os.path.isdir(outd)):
            # Nothing for now, perhaps should delete?
            self.logger.warning(
                "Output directory %s already exists, adding/updating files" % outd)
            pass
        elif (os.path.isfile(outd)):
            raise IIIFStaticError(
                "Can't write to directory %s: a file of that name exists" % outd)
        else:
            os.makedirs(outd)
        self.logger.debug("Output directory %s" % outd)

    def write_html(self, html_dir='/tmp', include_osd=False,
                   osd_width=500, osd_height=500):
        """Write HTML test page using OpenSeadragon for the tiles generated.

        Assumes that the generate(..) method has already been called to set up
        identifier etc. Parameters:
          html_dir - output directory for HTML files, will be created if it
                     does not already exist
          include_osd - true to include OpenSeadragon code
          osd_width - width of OpenSeadragon pane in pixels
          osd_height - height of OpenSeadragon pane in pixels
        """
        osd_config = self.get_osd_config(self.osd_version)
        osd_base = osd_config['base']
        osd_dir = osd_config['dir']  # relative to base
        osd_js = os.path.join(osd_dir, osd_config['js'])
        osd_images = os.path.join(osd_dir, osd_config['images'])
        if (os.path.isdir(html_dir)):
            # Exists, fine
            pass
        elif (os.path.isfile(html_dir)):
            raise IIIFStaticError(
                "Can't write to directory %s: a file of that name exists" % html_dir)
        else:
            os.makedirs(html_dir)
        self.logger.info("Writing HTML to %s" % (html_dir))
        with open(os.path.join(self.template_dir, 'static_osd.html'), 'r') as f:
            template = f.read()
        outfile = self.identifier + '.html'
        outpath = os.path.join(html_dir, outfile)
        with open(outpath, 'w') as f:
            info_json_uri = '/'.join([self.identifier, 'info.json'])
            if (self.prefix):
                info_json_uri = '/'.join([self.prefix, info_json_uri])
            d = dict(identifier=self.identifier,
                     api_version=self.api_version,
                     osd_version=self.osd_version,
                     osd_uri=osd_js,
                     osd_images_prefix=osd_images,
                     osd_height=osd_width,
                     osd_width=osd_height,
                     info_json_uri=info_json_uri)
            f.write(Template(template).safe_substitute(d))
            self.logger.info("%s / %s" % (html_dir, outfile))
        # Do we want to copy OSD in there too? If so, do it only if
        # we haven't already
        if (include_osd):
            if (self.copied_osd):
                self.logger.info("OpenSeadragon already copied")
            else:
                # Make directory, copy JavaScript and icons (from osd_images)
                osd_path = os.path.join(html_dir, osd_dir)
                if (not os.path.isdir(osd_path)):
                    os.makedirs(osd_path)
                shutil.copyfile(os.path.join(osd_base, osd_js),
                                os.path.join(html_dir, osd_js))
                self.logger.info("%s / %s" % (html_dir, osd_js))
                osd_images_path = os.path.join(html_dir, osd_images)
                if (os.path.isdir(osd_images_path)):
                    self.logger.warning(
                        "OpenSeadragon images directory (%s) already exists, skipping"
                        % osd_images_path)
                else:
                    shutil.copytree(os.path.join(osd_base, osd_images),
                                    osd_images_path)
                    self.logger.info("%s / %s/*" % (html_dir, osd_images))
                self.copied_osd = True  # don't try again for next img
